akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala

```diff
@@ -133,18 +134,32 @@ object TraversalBuilder {
    */
   private[impl] def initShape(shape: Shape): Unit = {
     // Initialize port IDs
-    val inIter = shape.inlets.iterator
-    var i = 0
-    while (inIter.hasNext) {
-      inIter.next.id = i
-      i += 1
+    val inlets = shape.inlets
+    if (inlets.nonEmpty) {
```

Comment:
You mean something smarter for calculating the size of List with 0 or 1 elements?
I don't think we can do that much better than what is already done for `List`, see `length` in `LinearSeqOptimized` (no allocations)